sdk/rust/oak/tests/handle_extract_inject.rs

```diff
@@ -0,0 +1,182 @@
+fn extract_struct() {
+    let mut message = TestMessage {
+        other_arbitrary_field: "Test".to_string(),
+        test_sender: Some(sender(42)),
+        test_receiver: Some(receiver(1337)),
+    };
+
+    let handles = extract_handles(&mut message);
+
+    assert_eq!(handles, vec![42, 1337]);
+    assert_eq!(
+        message,
+        TestMessage {
+            other_arbitrary_field: "Test".to_string(),
+            test_sender: Some(sender(0)),
+            test_receiver: Some(receiver(0)),
+        }
+    );
+}
+
+#[test]
+fn enum_extract_sender() {
+    let mut message = TestMessageWithEnum {
+        either: Some(test_message_with_enum::Either::EitherSender(sender(42))),
+    };
+
+    let handles = extract_handles(&mut message);
+
+    assert_eq!(handles, vec![42]);
+    assert_eq!(
+        message,
+        TestMessageWithEnum {
+            either: Some(test_message_with_enum::Either::EitherSender(sender(0))),
+        }
+    );
+}
+
+#[test]
+fn enum_extract_receiver() {
+    let mut message = TestMessageWithEnum {
+        either: Some(test_message_with_enum::Either::EitherReceiver(receiver(42))),
+    };
+
+    let handles = extract_handles(&mut message);
+
+    assert_eq!(handles, vec![42]);
+    assert_eq!(
+        message,
+        TestMessageWithEnum {
+            either: Some(test_message_with_enum::Either::EitherReceiver(receiver(0))),
+        }
+    );
+}
+
+#[test]
+fn enum_inject_sender() {
+    let mut message = TestMessageWithEnum {
+        either: Some(test_message_with_enum::Either::EitherSender(sender(0))),
+    };
+
+    inject_handles(&mut message, &[42]).unwrap();
+
+    assert_eq!(
+        message,
+        TestMessageWithEnum {
+            either: Some(test_message_with_enum::Either::EitherSender(sender(42))),
+        }
+    );
+}
+
+#[test]
+fn map_extract() {
+    use dummy_hash::DummyBuildHasher;
+    use std::collections::HashMap;
+    let mut map: HashMap<u64, Sender<()>, DummyBuildHasher> =
+        HashMap::with_hasher(DummyBuildHasher);
+    map.insert(1, sender(10));
+    map.insert(2, sender(20));
+    // DummyHasher should yield elements in reverse order.
+    assert_eq!(
+        map.values().cloned().collect::<Vec<Sender<()>>>(),
+        vec![sender(20), sender(10)]
+    );
+
+    let handles = extract_handles(&mut map);
+
+    // Even though the hashmap returns the values in reverse order, we expect the values to be
+    // extracted in the order of their keys.
+    assert_eq!(handles, vec![10, 20]);
+}
+
+#[test]
+fn map_inject() {
+    use dummy_hash::DummyBuildHasher;
+    use std::collections::HashMap;
+    let mut map: HashMap<u64, Sender<()>, DummyBuildHasher> =
+        HashMap::with_hasher(DummyBuildHasher);
+    map.insert(1, sender(0));
+    map.insert(2, sender(0));
+
+    inject_handles(&mut map, &[10, 20]).unwrap();
+
+    assert_eq!(map.get(&1).cloned(), Some(sender(10)));
+    assert_eq!(map.get(&2).cloned(), Some(sender(20)));
+}
+
+#[test]
+fn recursive_extract() {
```

Comment:
Would be useful to have some handles in the parent message before and after the child message, and have the child message repeated too. This would give us more confidence that the traversal order is well defined.

Reply:
Done.